flocker/node/agents/blockdevice.py

```diff
@@ -340,3 +504,113 @@ def get_device_path(self, blockdevice_id):
         )
         # May be None if the file hasn't been used for a loop device.
         return _device_for_path(volume_path)
+
+
+def _manifestation_from_volume(volume):
+    """
+    :param BlockDeviceVolume volume: The block device which has the
+        manifestation of a dataset.
+    :returns: A primary ``Manifestation`` of a ``Dataset`` with the same id as
+        the supplied ``BlockDeviceVolume``.
+    """
+    dataset = Dataset(
+        dataset_id=volume.dataset_id,
+        maximum_size=volume.size,
+    )
+    return Manifestation(dataset=dataset, primary=True)
+
+
+@implementer(IDeployer)
+class BlockDeviceDeployer(PRecord):
+    """
+    An ``IDeployer`` that operates on ``IBlockDeviceAPI`` providers.
+
+    :param unicode hostname: The IP address of the node that has this deployer.
+    :param IBlockDeviceAPI block_device_api: The block device API that will be
+        called upon to perform block device operations.
+    :ivar FilePath mountroot: The directory where block devices will be
+        mounted.
+    """
+    hostname = field(type=unicode, mandatory=True)
+    block_device_api = field(mandatory=True)
+    mountroot = field(type=FilePath, initial=FilePath(b"/flocker"))
+
+    def discover_local_state(self):
+        """
+        Find all block devices that are currently associated with this host and
+        return a ``NodeState`` containing only ``Manifestation`` instances and
+        their mount paths.
+        """
+        volumes = self.block_device_api.list_volumes()
+
+        manifestations = [_manifestation_from_volume(v)
+                          for v in volumes
+                          if v.host == self.hostname]
+
+        paths = {}
+        for manifestation in manifestations:
+            dataset_id = manifestation.dataset.dataset_id
+            mountpath = self._mountpath_for_manifestation(manifestation)
+            paths[dataset_id] = mountpath
+
+        state = NodeState(
+            hostname=self.hostname,
+            running=(),
```

Comment:
Woo, merge issues waiting to happen. I guess whoever merges second gets to fix this.